### PR TITLE
DNS encryption

### LINKED ISSUE #871: DNS over HTTPS
y/n option for routing DNS over HTTPS

[Dingo](https://github.com/pforemski/dingo) can do it with Google DNS, or we can go the route that [Cloudflare did](https://blog.cloudflare.com/dns-over-tls-for-openwrt/) with DNSmasq and Unbound.

### PR DESCRIPTION
**Pull request by jackivanov**
- [dnscrypt-proxy](https://github.com/jedisct1/dnscrypt-proxy) service for supported *BSDs and Ubuntu.
- AppArmor policies and cgroups limits
- CloudFlare DNS over HTTPS enabled by default for all installations
- Fixes #871

### CODE REVIEW

File: roles/dns_adblocking/tasks/ubuntu.yml [hunk 1]
```diff
@@ -7,13 +7,13 @@
     owner: root
     group: root
     mode: 0600
-  when: apparmor_enabled is defined and apparmor_enabled == true
+  when: apparmor_enabled|default(flase)|bool == true
```
Comment (jplock) [inline roles/dns_adblocking/tasks/ubuntu.yml]: `false`, same comment below

File: roles/vpn/tasks/ubuntu.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 - name: Ubuntu | Enforcing ipsec with apparmor
   shell: aa-enforce "{{ item }}"
-  when: apparmor_enabled is defined and apparmor_enabled == true
+  when: apparmor_enabled|default(flase)|bool == true
```
Comment (jplock) [inline roles/vpn/tasks/ubuntu.yml]: `false`

**Comment by dguido** [comment]
Let's put this in the default configuration, rather than behind the ad blocking role.

I assume that many users may want to change between Google and Cloudflare for DNS-over-HTTPS. Is the setting for that exposed in config.cfg?

**Comment by dguido** [comment]
You know what, let's switch the default to Cloudflare DNS. Cloudflare deletes a larger amount of data from their logs than Google does:

Compare:

* https://developers.google.com/speed/public-dns/privacy
* https://developers.cloudflare.com/1.1.1.1/commitment-to-privacy/privacy-policy/privacy-policy/

**Comment by dguido** [comment]
Yes, skip the y/n prompt. Let’s make this the default for everyone.

**Comment by dguido** [comment]
What does Unbound do? Do we need it or can we just use Dingo?

**Comment by jackivanov** [comment]
Unbound for CloudFlare. Dingo [doesn't support](https://github.com/pforemski/dingo/pull/41) CloudFlare.

**Comment by beachfork** [comment]
If Clouflare only would be an option, cloudflared (argo-tunnel) for DNS-over-HTTPS would be easy to integrate with dnsmasq.

https://developers.cloudflare.com/1.1.1.1/dns-over-https/cloudflared-proxy/

**Comment by jackivanov** [comment]
@beachfork Any benefits compared to Unbound? I don't see a binary for *BSDs

**Comment by beachfork** [comment]
Oh you're right, there's no *BSDs support for now. As far as I know there is DNS-over-TLS and DNS-over-HTTPS (DoH), cloudflared is for DoH. The Unbound method will get you DNS-over-TLS.

**Comment by jackivanov** [comment]
Right. I'd prefer DNS-over-TLS. At least it's RFCed. And should be faster in theory

**Comment by beachfork** [comment]
From benchmarks over the internet DNS-over-TLS seems to be faster! So you're right, Unbound is the universal route.

**Comment by dguido** [comment]
Performance should be similar between HTTPS and TLS for our setup. The AlgoVPN server is doing the lookup, and it will do it from a VPS datacenter over a fast, wired connection. I think it's fine to support _only_ DNS-over-HTTPS.

There is a draft RFC. It requires HTTP2 with strict cipher suites, so it's pretty good. https://www.ietf.org/id/draft-ietf-doh-dns-over-https-07.txt

Let's ditch the Unbound client. I am concerned that it adds too much complexity and it's another random pile of C that we have to stay concerned about. I'm much happier with a simple Go binary with Dingo.

**Comment by dguido** [comment]
Damn it. I just noticed that Dingo may have trouble with Cloudflare DNS. Let's use dnscrypt-proxy instead: https://github.com/jedisct1/dnscrypt-proxy

**Comment by ndtob** [comment]
Maybe it's too late but Stubby https://dnsprivacy.org/wiki/display/DP/About+Stubby is what I used for DNS-over-TLS local resolver. Just throwing it out there. Supports IPv4/6 etc.

**Comment by dguido** [comment]
Niceeee. Let me know when this is ready for review! I'll deploy it and test it.

**Comment by jackivanov** [comment]
Rebased

**Comment by dguido** [comment]
Ehhh never mind. Let’s get it out there quickly.

**Pull request merged**